### ISSUE TITLE
Mac OSX compile attempt

### ISSUE THREAD

**Issue by g-cole** [ComputerNerd/Retro-Graphics-Toolkit#8]
First off, thank you for this program, even if it's not working for me yet. Most similar programs are closed-source Windows one-offs, that are not expandable or supportable; I can't even attempt to compile them or am forced to use a VM for a dinky pixel art program. I am new to development on a mac and I have no desire to use xcode (which may be an issue here?). I use a mac because of the battery life and they're pretty much the standard at my office, but I shy away from apple otherwise.

Trying to build on Mac OSX 10.11.6
- copy lua 5.3.4 into lua directory
- edit makefile src/lua for "CC=clang++" and "-DLUA_COMPAT_5_2" removal
- fltk 1.3.4 is installed via homebrew

make
```
	clang: warning: optimization flag '-fuse-linker-plugin' is not supported
	clang: warning: argument unused during compilation: '-s'
	In file included from project.cpp:17:
	./includes.h:25:10: fatal error: 'endian.h' file not found
```

I found a portable endian.h file online that takes care of some errors, like htobe16 definition. It allows further compilation with many warnings...

	main.cpp:59:24: error: use of undeclared identifier 'get_current_dir_name'
        const char*olddirname=get_current_dir_name();

I barely know what I'm doing here, just trying to get a successful compilation. After a few minutes research, I find that PATH_MAX & getcwd might work, but it is probably not really correct or at least not recommended.

edit main.cpp
```
	#ifdef _WIN32
		char olddirname[MAX_PATH];
		getcwd(olddirname,sizeof(olddirname));
	#elif __APPLE__
		char olddirname[PATH_MAX];
		getcwd(olddirname,sizeof(olddirname));
	#else
		const char*olddirname=get_current_dir_name();
```

Progress though.

`runlua.cpp:80:10: fatal error: 'posix.h' file not found`

 Any suggestions? I feel like I could just be missing something fundamental...

**Comment by ComputerNerd** [comment]
The reason you got the posix.h error is because I forgot to add the file to the repository. I added the file. Try running git pull and then try again.

**Comment by g-cole** [comment]
It does not like the makefile changes.
```
Makefile:5: Makefile.common: No such file or directory
make: *** No rule to make target 'Makefile.common'.  Stop.
```
So I reverted to the old makefile (with -Ilua/src).

I again added the portable endian.h from [https://gist.github.com/panzi/6856583](https://gist.github.com/panzi/6856583) and changed 'get_current_dir_name' like in my first post.

Latest error... (tried fltk 1.3.4 and 1.3.3)
`autogeneratedBindings/FLTK_Fl_Box.cpp:27:18: error: use of undeclared identifier 'free'`

Thanks for the reply! I'll try to debug more soon, but for me that involves "learning make and c++ workflows". Not exactly my area.

**Comment by ComputerNerd** [comment]
I fixed the Makefile.common issue. It was caused by forgetting to include it in the repository.

I think I fixed the free issue by including stdlib.h which contains the needed function prototype for free.

I also noticed a bug in the new clang make files. I have fixed that as well.

**Comment by g-cole** [comment]
Things seem much smoother with this last update.

`luaposix/strlcpy.h:43:15: error: static declaration of 'strlcpy' follows non-static declaration`
removed "static", also had to add -D_FORTIFY_SOURCE=0 to CFLAGS or I get "expected parameter declarator" and "conflicting types" errors for strlcpy.

```
ld: library not found for -lXft
clang: error: linker command failed with exit code 1
```
I noticed OSX is missing the following from LDFLAGS: -lXft -lXext -lXinerama -lX11
I installed freetype via homebrew, but I guess newer versions of OSX don't support regular X stuff anymore. I'll look into it, see if there is an easy solution, and try again soon. I'll install XQuartz if needed to get a compile, but that feels excessive.

Thanks again for your quick replies!

**Comment by ComputerNerd** [comment]
Those are all there because of FLTK. For Mac OS X FLTK uses Quartz and Cocoa. You may be able to remove all those libraries and replace it with Quartz. I looked at what the FLTK project is doing and they just use "-framework Cocoa" (without the quotes) in the LDFLAGS. Let me know if this approach works and if so I will create a separate Mac OS X Makefile that uses those libraries.

**Comment by g-cole** [comment]
Alright, I started with a fresh install of OS X 10.11.6 so I know there is nothing conflicting. Here are the exact steps I’ve taken. I apologize for the redundancy in regard to my previous posts... for posterity I guess.

1. Install homebrew, which installs the mac command-line/build tools by default
2. Download latest lua source, stick in project dir, and change it’s makefile options (PLAT= macosx, CC= clang++, remove “-DLUA_COMPAT_5_2”)
3. Put portable “[endian.h](https://gist.github.com/panzi/6856583)” into project dir
4. Download fltk-1.3.4-1 source into the project folder and ran ./configure and make
5. Added “-Ifltk-1.3.4-1/ -Ifltk-1.3.4-1/zlib -Ifltk-1.3.4-1/png -D_FORTIFY_SOURCE=0” to CFLAGS 
6. Removed “static” from luaposix/strlcpy.c & .h
7. Removed the strlcpy.c include from luaposix/ext/posix/sys/socket.cpp for duplicate symbol error
8. Removed “-L/usr/lib/fltk/ -lfltk_images -lfltk -lpng -ljpeg -lXft -lXext -lXinerama -lX11” from LDFLAGS and replaced with “-framework Cocoa -Lfltk-1.3.4-1/lib”

Errors:
```
undef: __Z15luaL_fileresultP9lua_StateiPKc
undef: __Z15luaL_checkudataP9lua_StateiPKc
undef: __Z15lua_tocfunctionP9lua_Statei
undef: __Z14luaL_checktypeP9lua_Stateii
undef: __ZN13Fl_Tree_Prefs9closeiconEP8Fl_Image
…etc…
```
large list of undefs
```
Undefined symbols for architecture x86_64:
  "luaL_fileresult(lua_State*, int, char const*)", referenced from:
      stdio_fclose(lua_State*) in lto.o
  "luaL_checkudata(lua_State*, int, char const*)", referenced from:
      Pfileno(lua_State*) in lto.o
      stdio_fclose(lua_State*) in lto.o
  "lua_tocfunction(lua_State*, int)", referenced from:
      Psignal(lua_State*) in lto.o
      ...
  "Fl_Tree_Prefs::closeicon(Fl_Image*)", referenced from:
      Fl_Tree_Prefs_closeicon(lua_State*) in lto.o
…etc…
```
large list of lua and Fl symbol errors, ending with
```
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [RetroGraphicsToolkit] Error 1
```

**Comment by ComputerNerd** [comment]
I think the problem may be that you are not using make.clang. I see mentions of lto.o however for clang I don't have LTO enable in make.clang. Other than that it looks like you did everything right. To use make.clang run make -f make.clang. Let me know if using make.clang fixes your problem and if not I will try and use Mac OS X myself following the steps in your summary.

I also pushed a change that might fix the strlcpy issue for you without having to patch it every-time.

**Comment by g-cole** [comment]
With make -f make.clang and the latest pull, there are no more fltk Fl symbol errors or list of undefs, but there are still the lua symbol errors.
```
Undefined symbols for architecture x86_64:
  "luaL_error(lua_State*, char const*, ...)", referenced from:
      runLuaFunc(lua_State*, unsigned int, unsigned int) in runlua.o
      settings__set_(lua_State*) in runlua.o
      dub::checksdata_n(lua_State*, int, char const*, bool) in dub.o
      levDat__set_(lua_State*) in level_levDat.o
      levelInfo__set_(lua_State*) in level_levelInfo.o
      levobjDat__set_(lua_State*) in level_levobjDat.o
  "luaL_where(lua_State*, int)", referenced from:
      dub::error(lua_State*) in dub.o
  "lua_concat(lua_State*, int)", referenced from:
      lz_filter_impl(lua_State*, int (*)(z_stream_s*, int), int (*)(z_stream_s*), char*) in lua_zlib.o
...etcetc...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [RetroGraphicsToolkit] Error 1
```
I tried a few different make options for lua and researched a bit, but it seems like this has some pretty specific lua requirements. The resources I found with similar errors suggested compiling lua with compatibility flags, but your compile document is pretty clear. I wish I could be more helpful, but for what it's worth, I'm learning a lot...

**Comment by ComputerNerd** [comment]
The problem was that the Makefile in lua/src/ was overriding CC = clang++. The reason I did not catch this issue is because the overriding is only for Mac OS X users who build Lua using make macosx.
The solution is to edit lua/src/Makefile and look for:

    $(MAKE) $(ALL) SYSCFLAGS="-DLUA_USE_MACOSX" SYSLIBS="-lreadline" CC=cc

and replace it with:

    $(MAKE) $(ALL) SYSCFLAGS="-DLUA_USE_MACOSX" SYSLIBS="-lreadline"

In other words remove CC=cc from the end of the line. 

After making that change and saving it run:

    make clean && make macosx

When compiling code with a C++ compiler the names are mangled. Because Lua was accidentally built with a C compiler instead of a C++ compiler it could not find those functions because their names were not mangled like Clang++ expected.

**Comment by g-cole** [comment]
That was it! Fantastic! It successfully compiled and everything looks correct, though I haven't played with it much yet. That was rather painless once you solved those Mac quirks. Now I'll finally get to figure out how to use this program and scratch my nostalgia itch.

I'll close this issue, but please let me know if I can assist with testing compile instructions, documentation, or in some sort of binary/.app/homebrew release... or I'll just push changes via git now that I know there is nothing wrong with my system.

Thanks again for your help with this!
